### ISSUE TITLE
[FEATURE]: Add native backend resolver for Hugging Face transformers

### ISSUE THREAD

**Issue by Khanz9664** [Khanz9664/TrustLens#123]
### Summary

**Problem Statement**
Currently, TrustLens offers a seamless, zero-friction auditing experience for traditional ML models via native backend resolvers (`sklearn.py`, `xgboost.py`). However, as practitioners increasingly rely on fine-tuned LLMs and BERT variants for text classification, they are blocked. To audit a Hugging Face model today, a user must manually extract raw logits, apply softmax, align string labels, and convert everything to NumPy arrays before calling `analyze()`. This introduces massive friction for the NLP community.

**Proposed Solution**
We need to introduce a native backend resolver for the Hugging Face ecosystem. 
Because TrustLens uses a decoupled `PredictionBundle` architecture, we can easily plug this in without changing the core analysis engine.

### Where to Edit
1. **`trustlens/backends/huggingface.py`** (New File): Implement the resolver logic.
2. **`trustlens/backends/registry.py`**: Register the new framework.
3. **`tests/test_huggingface_backend.py`** (New File): Add basic unit tests.
4. **`pyproject.toml`**: Add optional `nlp` dependency group.

### Proposed Implementation Steps

**1. Create the Resolver (`trustlens/backends/huggingface.py`)**
Create a `resolve()` function that matches the signature in `sklearn.py`. 
* Check if the `model` is a `transformers.Pipeline` (specifically `TextClassificationPipeline`).
* If `y_prob` is not provided, pass the input strings `X` through the pipeline: `model(X, top_k=None)`.
* Parse the returned list of dictionaries (which contain `{'label': ..., 'score': ...}`).
* Construct a 2D `np.ndarray` of probabilities (`y_prob`).
* Construct a 1D `np.ndarray` of predicted integer indices (`y_pred`).
* Return a `PredictionBundle(y_pred=y_pred, y_prob=y_prob, framework="huggingface", class_labels=...)`.

**2. Update the Registry (`trustlens/backends/registry.py`)**
* Add `"transformers": "huggingface"` to the `FRAMEWORK_MAPPING` dictionary.
* Update `IMPLEMENTED_RESOLVERS` to include `"huggingface"`.
* In `get_resolver()`, add the routing logic:
  ```python
  if detected == "huggingface":
      from trustlens.backends import huggingface
      return huggingface.resolve
  ```

**3. Optional Dependencies (`pyproject.toml`)**
Because the `transformers` and `torch` libraries are massive, this integration **must** be implemented as an optional dependency to avoid bloating the core package for tabular users. 
Add an optional dependency group:
```toml
[project.optional-dependencies]
nlp = ["transformers", "torch"]
```
*(Ensure all imports inside `huggingface.py` are wrapped in `try/except ImportError` blocks).*

**Alternatives Considered**
We could maintain the status quo and require NLP users to pass `framework="manual"` alongside pre-computed `y_prob` arrays. However, because Hugging Face is the undisputed industry standard for NLP, missing native support limits TrustLens's adoption. 

### How to Verify
Write a mock test in `tests/test_huggingface_backend.py`. You do not need to download a massive model; you can use the tiny `hf-internal-testing/tiny-random-DistilBertForSequenceClassification` or simply mock the `Pipeline` object to ensure it returns the correct NumPy arrays.

```bash
pytest tests/test_huggingface_backend.py
```

### 🤝 Need Help?
This is a fantastic issue for a contributor looking to dive into the core architecture of TrustLens while working with modern NLP tools! If you have any questions about the `PredictionBundle` contract or Hugging Face pipelines, feel free to comment on this issue or join our Discussions.

**Comment by AaronProbha18** [comment]
Hi! @Khanz9664 
I'd like to work on this issue.

## Plan

The goal is to add support for Hugging Face text-classification pipelines so users can pass a `TextClassificationPipeline` directly to `analyze()` without manually converting predictions into NumPy arrays. The implementation follows the existing backend architecture by introducing a dedicated resolver rather than extending the manual path.

I'll start by confirming the runtime output format of `TextClassificationPipeline` and the canonical label ordering exposed through the model configuration, since both are important for constructing a stable probability matrix. From there, I'll add the new backend resolver, register it with the framework detection and dispatch logic, update the optional dependency configuration, and finish with tests and documentation. The resolver will intentionally remain scoped to text-classification pipelines, while preserving the existing manual override behavior and raising `UnsupportedModelError` for unsupported pipeline types.

## Changes

- `trustlens/backends/huggingface.py` will introduce the new `resolve()` implementation, handling pipeline execution, probability extraction, prediction generation, metadata population, and `PredictionBundle` creation while following the same contract as the existing sklearn backend.
- `trustlens/backends/registry.py` will be updated to detect `transformers` models, register the new backend, and lazily import the resolver using the existing backend dispatch pattern.
- `trustlens/backends/types.py` will remain unchanged, with the new resolver reusing the existing `PredictionBundle` and `UnsupportedModelError` contracts.
- `trustlens/backends/sklearn.py` will serve as the implementation reference to keep resolver signatures, metadata, and error handling consistent.
- `pyproject.toml` will add a standalone `transformers` optional dependency, include it in the `full` extra, and continue relying on the existing `torch` extra instead of duplicating dependencies.
- `tests/test_backend_huggingface.py` will add coverage for framework detection, resolver behavior, end-to-end `analyze()` integration, manual override support, and rejection of unsupported pipeline types, following the conventions used in `tests/test_backend_xgboost.py`.

**Comment by Khanz9664** [comment]
@AaronProbha18 

Thanks for the detailed implementation plan! It looks like you've taken the time to understand the existing backend architecture and the separation between the resolver layer and the core analysis pipeline, which is exactly the right starting point.

I like that you're keeping the initial scope limited to `TextClassificationPipeline`. Keeping the first implementation focused will make it much easier to review and extend later.

A couple of implementation notes before you begin:

* Please keep the resolver framework-agnostic from the perspective of the core engine. The resolver's responsibility should be limited to producing a valid `PredictionBundle`; avoid introducing Hugging Face-specific logic outside the backend layer.
* I'd recommend relying on the model's configured label mapping (`id2label` / `label2id`) to construct a deterministic class ordering rather than depending on the order returned by the pipeline output.
* Please continue following the existing lazy-import pattern used by the other backends so users who don't install the NLP extras aren't affected.
* For testing, prefer lightweight mocks where possible. An end-to-end integration test using a tiny Hugging Face model is great, but the majority of the test suite should remain fast and deterministic without requiring network access.
* Please also verify that manual overrides (`framework="manual"` and user-supplied predictions) continue to behave exactly as they do today. Backward compatibility is important here.

Other than that, the proposed scope looks good to me. Feel free to proceed, and if you run into any questions around the resolver contract or registry integration, don't hesitate to ask before going too far down an implementation path. Looking forward to reviewing the PR!
